File: app/src/main/java/br/com/jonathanzanella/myexpenses/account/AccountView.java
package br.com.jonathanzanella.myexpenses.account;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.views.BaseView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class AccountView extends BaseView {
	private static final int REQUEST_ADD_ACCOUNT = 1003;

	@BindView(R.id.view_accounts_list)
	RecyclerView accounts;

	private AccountAdapter adapter;

	public AccountView(Context context) {
		super(context);
	}

	public AccountView(Context context, AttributeSet attrs) {
		super(context, attrs);
	}

	public AccountView(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
	}

	@Override
	protected void init() {
		inflate(getContext(), R.layout.view_accounts, this);
		ButterKnife.bind(this);

		adapter = new AccountAdapter();

		accounts.setAdapter(adapter);
		accounts.setLayoutManager(new GridLayoutManager(getContext(), 2));
		accounts.setItemAnimator(new DefaultItemAnimator());
	}

	@OnClick(R.id.view_accounts_fab)
	void onFab() {
		Context ctx = getContext();
		Intent i = new Intent(getContext(), EditAccountActivity.class);
		if(ctx instanceof Activity) {
			((Activity) ctx).startActivityForResult(i, REQUEST_ADD_ACCOUNT);
		} else {
			ctx.startActivity(i);
		}
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);

		switch (requestCode) {
			case REQUEST_ADD_ACCOUNT:
				if(resultCode == Activity.RESULT_OK)
					adapter.refreshData();
				break;
		}
	}

	@Override
	public void refreshData() {
		super.refreshData();

		adapter.refreshData();
		adapter.notifyDataSetChanged();
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/account/AccountView.kt
package br.com.jonathanzanella.myexpenses.account

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.widget.FrameLayout
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.views.FilterableView
import br.com.jonathanzanella.myexpenses.views.TabableView
import br.com.jonathanzanella.myexpenses.views.anko.applyTemplateViewStyles
import br.com.jonathanzanella.myexpenses.views.anko.recyclerView
import org.jetbrains.anko.*
import org.jetbrains.anko.design.floatingActionButton
import javax.inject.Inject

class AccountView@JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr), FilterableView, TabableView {
    override var filter: String = ""

    @Inject
    lateinit var adapter: AccountAdapter
    private val ui = AccountViewUI()

    init {
        App.getAppComponent().inject(this)
        addView(ui.createView(AnkoContext.Companion.create(context, this)))

        ui.accounts.adapter = adapter
        ui.accounts.layoutManager = GridLayoutManager(context, 2)
        ui.accounts.itemAnimator = DefaultItemAnimator()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        adapter.setFormat(AccountAdapter.Format.LIST)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        adapter.onDestroy()
    }

    companion object {
        val REQUEST_ADD_ACCOUNT = 1003
    }
}

class AccountViewUI: AnkoComponent<AccountView> {
    lateinit var accounts : RecyclerView

    override fun createView(ui: AnkoContext<AccountView>) = with(ui) {
        frameLayout {
            accounts = recyclerView { id = R.id.view_accounts_list }
            floatingActionButton {
                id = R.id.view_accounts_fab
                onClick {
                    val i = Intent(context, EditAccountActivity::class.java)
                    if (ctx is Activity) {
                        (ctx as Activity).startActivityForResult(i, AccountView.REQUEST_ADD_ACCOUNT)
                    } else {
                        ctx.startActivity(i)
                    }
                }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
